File: WaterMeter.cs
// WaterMeter.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;

class Entry
{
    [JsonPropertyName("date")] public string Date { get; set; }
    [JsonPropertyName("value")] public double Value { get; set; }
    [JsonPropertyName("unit")] public string Unit { get; set; }
}

class WaterMeter
{
    private List<Entry> data = new List<Entry>();
    private readonly string dataFile = "water_data.json";
    private readonly JsonSerializerOptions options = new JsonSerializerOptions { WriteIndented = true };

    public WaterMeter() => Load();

    private void Load()
    {
        if (!File.Exists(dataFile)) return;
        string json = File.ReadAllText(dataFile);
        data = JsonSerializer.Deserialize<List<Entry>>(json) ?? new List<Entry>();
    }

    private void Save()
    {
        string json = JsonSerializer.Serialize(data, options);
        File.WriteAllText(dataFile, json);
    }

    public void Add(double value, string unit = "L", string date = null)
    {
        if (string.IsNullOrEmpty(date)) date = DateTime.Now.ToString("yyyy-MM-dd");
        if (string.IsNullOrEmpty(unit)) unit = "L";
        data.Add(new Entry { Date = date, Value = value, Unit = unit });
        Save();
        Console.WriteLine($"✅ Added: {value} {unit} on {date}");
    }

    public void List()
    {
        if (!data.Any()) { Console.WriteLine("No data."); return; }
        Console.WriteLine("\n📋 All readings:");
        foreach (var e in data)
            Console.WriteLine($"  {e.Date}: {e.Value} {e.Unit}");
    }

    public void Stats(int days = 7, string unit = null)
    {
        if (!data.Any()) { Console.WriteLine("No data."); return; }
        var cutoff = DateTime.Now.AddDays(-days);
        var filtered = data.Where(e => DateTime.Parse(e.Date) >= cutoff).ToList();
        if (!filtered.Any())
        {
            Console.WriteLine($"No data in the last {days} days.");
            return;
        }
        string targetUnit = unit ?? "L";
        double total = 0;
        foreach (var e in filtered)
        {
            double val = e.Value;
            if (e.Unit == "m3" && targetUnit == "L") val *= 1000;
            else if (e.Unit == "L" && targetUnit == "m3") val /= 1000;
            total += val;
        }
        double avg = total / filtered.Count;
        Console.WriteLine($"\n📊 Statistics (last {days} days) – unit: {targetUnit}");
        Console.WriteLine($"  Total: {total:F2} {targetUnit}");
        Console.WriteLine($"  Average per day: {avg:F2} {targetUnit}");
    }

    public void Reset()
    {
        data.Clear();
        Save();
        Console.WriteLine("All data cleared.");
    }

    static void Main(string[] args)
    {
        if (args.Length < 1)
        {
            Console.WriteLine("Usage: WaterMeter <command> [options]");
            return;
        }
        var app = new WaterMeter();
        string cmd = args[0];
        switch (cmd)
        {
            case "add":
                double val = double.Parse(args[1]);
                string unit = "L";
                string date = null;
                for (int i=2; i<args.Length; i++)
                {
                    if (args[i] == "--unit" && i+1 < args.Length) unit = args[++i];
                    if (args[i] == "--date" && i+1 < args.Length) date = args[++i];
                }
                app.Add(val, unit, date);
                break;
            case "list":
                app.List();
                break;
            case "stats":
                int days = 7;
                string u = null;
                for (int i=1; i<args.Length; i++)
                {
                    if (args[i] == "--days" && i+1 < args.Length) days = int.Parse(args[++i]);
                    if (args[i] == "--unit" && i+1 < args.Length) u = args[++i];
                }
                app.Stats(days, u);
                break;
            case "reset":
                app.Reset();
                break;
            default:
                Console.WriteLine("Unknown command.");
                break;
        }
    }
}

File: water_meter.cpp
// water_meter.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <ctime>
#include <iomanip>
#include <sstream>
#include <nlohmann/json.hpp>

using namespace std;
using json = nlohmann::json;

struct Entry {
    string date;
    double value;
    string unit;
};

class WaterMeter {
private:
    vector<Entry> data;
    string dataFile = "water_data.json";

    string currentDate() {
        time_t t = time(nullptr);
        tm* now = localtime(&t);
        char buf[11];
        strftime(buf, sizeof(buf), "%Y-%m-%d", now);
        return string(buf);
    }

    void load() {
        ifstream f(dataFile);
        if (!f.is_open()) return;
        json j;
        f >> j;
        for (auto& item : j) {
            Entry e;
            e.date = item["date"];
            e.value = item["value"];
            e.unit = item["unit"];
            data.push_back(e);
        }
    }

    void save() {
        json j = json::array();
        for (auto& e : data) {
            j.push_back({{"date", e.date}, {"value", e.value}, {"unit", e.unit}});
        }
        ofstream f(dataFile);
        f << setw(2) << j << endl;
    }

public:
    WaterMeter() { load(); }

    void add(double value, const string& unit, const string& date) {
        string d = date.empty() ? currentDate() : date;
        string u = unit.empty() ? "L" : unit;
        data.push_back({d, value, u});
        save();
        cout << "✅ Added: " << value << " " << u << " on " << d << endl;
    }

    void list() {
        if (data.empty()) { cout << "No data." << endl; return; }
        cout << "\n📋 All readings:" << endl;
        for (auto& e : data) {
            cout << "  " << e.date << ": " << e.value << " " << e.unit << endl;
        }
    }

    void stats(int days, const string& unit) {
        if (data.empty()) { cout << "No data." << endl; return; }
        time_t now = time(nullptr);
        tm* now_tm = localtime(&now);
        time_t cutoff = now - days * 24 * 3600;
        vector<Entry> filtered;
        for (auto& e : data) {
            struct tm tm = {};
            strptime(e.date.c_str(), "%Y-%m-%d", &tm);
            time_t t = mktime(&tm);
            if (t >= cutoff) filtered.push_back(e);
        }
        if (filtered.empty()) {
            cout << "No data in the last " << days << " days." << endl;
            return;
        }
        string targetUnit = unit.empty() ? "L" : unit;
        double total = 0;
        for (auto& e : filtered) {
            double val = e.value;
            if (e.unit == "m3" && targetUnit == "L") val *= 1000;
            else if (e.unit == "L" && targetUnit == "m3") val /= 1000;
            total += val;
        }
        double avg = total / filtered.size();
        cout << "\n📊 Statistics (last " << days << " days) – unit: " << targetUnit << endl;
        cout << "  Total: " << fixed << setprecision(2) << total << " " << targetUnit << endl;
        cout << "  Average per day: " << fixed << setprecision(2) << avg << " " << targetUnit << endl;
    }

    void reset() {
        data.clear();
        save();
        cout << "All data cleared." << endl;
    }
};

int main(int argc, char* argv[]) {
    if (argc < 2) {
        cerr << "Usage: water_meter <command> [options]" << endl;
        return 1;
    }
    WaterMeter app;
    string cmd = argv[1];
    if (cmd == "add") {
        if (argc < 3) { cerr << "add <value> [unit] [--date YYYY-MM-DD]" << endl; return 1; }
        double value = stod(argv[2]);
        string unit = "L";
        string date;
        for (int i=3; i<argc; i++) {
            if (string(argv[i]) == "--date" && i+1 < argc) {
                date = argv[++i];
            } else if (string(argv[i]) != "--date") {
                unit = argv[i];
            }
        }
        app.add(value, unit, date);
    } else if (cmd == "list") {
        app.list();
    } else if (cmd == "stats") {
        int days = 7;
        string unit;
        for (int i=2; i<argc; i++) {
            if (string(argv[i]) == "--days" && i+1 < argc) {
                days = stoi(argv[++i]);
            }
            if (string(argv[i]) == "--unit" && i+1 < argc) {
                unit = argv[++i];
            }
        }
        app.stats(days, unit);
    } else if (cmd == "reset") {
        app.reset();
    } else {
        cerr << "Unknown command. Use add, list, stats, reset." << endl;
        return 1;
    }
    return 0;
}
